homekit: add native HeaterCooler accessory

nitaybz opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Adds a native Heater Cooler accessory to the HomeKit integration. Air conditioners and heat pumps get one tile with the mode, target temperature, cooling and heating thresholds, fan speed, and swing, instead of a Thermostat next to a separate Fan. When an integration does not report `hvac_action`, the current heating or cooling state is derived from the target and current temperature; when the entity has an automatic fan mode, the fan is exposed through a linked fan service so auto stays selectable; humidity that is only reported for display is shown through a linked humidity sensor; the temperature sliders follow the device's capabilities, so a cool only unit gets a single cooling slider.

Nothing changes for existing setups, so this is not a breaking change. Entities already bridged keep their Thermostat, and users switch them to the Heater Cooler at their own pace from the bridge options. Only climate entities that have never been bridged pick the Heater Cooler automatically, and only when they expose two or more predefined fan speeds like low and high, or a swing mode that can be turned off; timing fan modes like auto, on, off, and circulate do not count, so central thermostats such as ecobee, Nest, and Honeywell keep the Thermostat. Entities with a humidity setpoint stay on the Thermostat unless a type is set explicitly, since the Heater Cooler service cannot control humidity.

The accessory type can be changed at any time; UI bridges get a new step in the options flow with an automatic, thermostat, or heater cooler choice per climate entity, and YAML bridges use `type` in `entity_config`, mirroring the existing option for switch and fan entities. Docs in home-assistant/home-assistant.io#46577.

The fan, swing, and temperature handling is shared with the Thermostat through `climate_base` and `climate_util`, so the two accessories cannot drift apart.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #106028
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#46577
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies a diff between library versions and ideally a link to the changelog/release notes is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

bdraco reviewed (changes requested):
> * [ ]  **Tests added** – (unit tests will be added in a follow-up PR).

HomeKit changes require 100% coverage. They need to be included in the same PR the changes are made.

This is a significant breaking change. It needs a breaking change section and instruction on what do for users who are affected.

nitaybz commented:
@bdraco  I've added the "Braking Changes" section to the pull request and 100% test coverage

nitaybz commented:
@bdraco - I've implemented all the changes you requested:

Removed all debouncing logic and switched to a single service-level setter callback (serv.setter_callback = self._set_chars) that handles all characteristic changes at once. This eliminates the race conditions and hard-to-find bugs you mentioned.

Cleaned up all generated comments, improved error handling to catch specific exceptions only, aligned code structure with Thermostat patterns, and verified 100% test coverage. All pre-commit checks pass.

Ready for review.

nitaybz commented:
Hey @bdraco , it's been a while. would love to get an update, if i need to change anything further.

thanks :)

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,590 @@
+
+        # Check if entity supports auto or heat_cool modes
+        self._supports_auto = HVACMode.AUTO in hvac_modes
+        self._supports_heat_cool = HVACMode.HEAT_COOL in hvac_modes
```
This doesn't look used outside of init

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,590 @@
+        if not self.ordered_fan_speeds:
+            return
+
+        if 0 < speed <= 100:
```
Should use an early return pattern here

nitaybz commented:
Of course, I will be patient and would love to help any way I can to integrate this patch.

I'll be waiting for your queue 👍🏼

bdraco commented:
I was hoping to work on this next week but I just ended up having to travel next week so I will instead try to find time for this over the holidays at the end of the year when my day job isn't eating up my time

nitaybz commented:
Hey @bdraco! hope the hoilday went well and fun.
I was wondering if you found the time to work on it a little bit and if you think we will get to see this PR happening anytime soon.

thanks

semyonchetvertnyh commented:
Seems like it has to be added to Home Assistant because not every `climate` entity is a thermostat. Could I help somehow?

bdraco commented:
7 hours in on this today and there is still more cleanup to do. I'll hopefully have some time next month to look at it some more, but I've got to call it for today as I'm just too tired to continue and there is still a lot of code to check. The bot helps but it doesn't get everything and HomeKit has a lot of quirks

bdraco reviewed (dismissed):
This still needs a lot of work.

We need a documented discriminator so something like ecobee, nexia land in Thermostat and split units/portables land in HeaterCooler otherwise merging this PR will make a lot of people very unhappy as their Thermostats become HeaterCooler

bdraco committed:
Add native HomeKit HeaterCooler accessory

Climate entities that expose predefined fan speeds or a swing mode are now
exposed as a HomeKit Heater Cooler accessory instead of a Thermostat plus a
linked Fan, with the target temperature, cooling and heating thresholds,
rotation speed, and swing mode on one tile; a reported current humidity is
surfaced through a linked humidity sensor. Entities that control a humidity
setpoint, or expose no predefined fan speeds or swing, stay on the Thermostat.

Shared fan, swing, and temperature handling is factored into climate_base and
climate_util so the Thermostat and Heater Cooler behave identically. When an
integration omits hvac_action, the current heating or cooling state is derived
from the target and current temperature.

bdraco committed:
Harden off-only fallback and parametrize the temperature selection test

bdraco committed:
Require two or more predefined fan speeds to route to HeaterCooler

bdraco committed:
Clarify that timing fan modes are not predefined speeds

bdraco committed:
Test the zero temperature bound helper and document humidity topology

bdraco committed:
Guard against non-string fan modes in the fan speed helper

bdraco committed:
Share the case-insensitive mode map across the fan and swing helpers

bdraco committed:
Allow overriding the climate accessory type via entity config

bdraco committed:
Activate the displayed mode on power-on and scope single-setpoint writes

bdraco committed:
Keep the target mode in sync when off and document the action mapping

bdraco committed:
Explain why drying and fan actions map to Cooling

bdraco committed:
Merge branch 'dev' into heatercooler-support

bdraco commented:
> I'm not sure if @bluetoothbot reads these comments, but a suggestion for how to handle fan and dehumidifier: rather than having these map to states of the HeaterCooler accessory which don't correspond to what is actually happening, expose additional services in the accessory of type `HMServiceTypeFan` and `HMServiceTypeHumidifierDehumidifier` to control these functions. I've seen this pattern in Homebridge.

Let's not add more complexity to this PR. Its already very large, and adding more to it will make it unmergable.

bdraco committed:
Merge branch 'dev' into heatercooler-support

bdraco committed:
Restore the swing helper test dropped in the merge

bdraco committed:
Expose auto fan through a shared linked fan service

bdraco committed:
Address review feedback on batched writes and test hygiene

bdraco committed:
Expose only the capable temperature thresholds

bdraco committed:
Document the fan auto invariant and tighten the routing comment

bdraco committed:
Rename heater cooler module to match accessory module convention

bdraco committed:
Trim the climate routing comments

bdraco committed:
Merge branch 'dev' into heatercooler-support

bdraco committed:
Route climate to HeaterCooler automatically only for new entities

bdraco committed:
Offer HeaterCooler migration through a repair issue

bdraco committed:
Let the options flow choose the climate accessory type

bdraco committed:
Persist the automatic HeaterCooler choice so it survives restarts

bdraco committed:
Simplify the repairs flow and options flow routing

bdraco committed:
Address review feedback

Offer the climate options step when the whole domain is included,
resync repair issues when an accessory reloads, derive auto mode
action from a single setpoint, avoid the double cast in the fix
flow factory, and align the options copy with the routing behavior

bdraco committed:
Simplify climate type resolution and reconcile explicit choices

The resolver now owns candidate bookkeeping and records explicit
types in aid storage so switching back to automatic keeps the
accessory the entity already uses, accessory mode reloads resync
repair issues, bridge reloads only resync when a climate entity
was reloaded, and tests use the production id helpers

bdraco committed:
Address Copilot feedback on routing durability and auto mode writes

Store the HeaterCooler choice by the same stable identity as the aid
allocation so renames keep it, drop the stored choice when an entity
gains a humidity setpoint, treat auto like heat cool when selecting
a single setpoint threshold, and only remember modes the HomeKit
target can restore

bdraco committed:
Simplify stable identity lookups and reload issue syncing

One helper enumerates the storage keys an entity resolves to,
accessory removal ends repair candidacy, reload paths resync issues
unconditionally, the range mode pair is a shared constant, and the
dead branches around the target mode mapping are gone

bdraco committed:
Address review feedback on accessory mode routing and rejected writes

A brand new accessory mode pairing now routes automatically like a
new bridge entity, rejected Active and target mode writes put the
characteristic back so HomeKit keeps showing the real state, the
migration text no longer promises controls the device lacks, and an
explicit type wins over the humidity safeguard, now tested

bdraco commented on homeassistant/components/homekit/aidmanager.py:
```diff
@@ -84,6 +86,23 @@ async def async_initialize(self) -> None:
+        if heater_cooler == (entity_id in entities):
+            return
+        if heater_cooler:
+            entities.add(entity_id)
```
e32598aba13

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,563 @@
+        # an arbitrary one.
+        self._last_known_mode: HVACMode
+        if current_mode and current_mode != HVACMode.OFF:
+            self._last_known_mode = current_mode
```
e32598aba13

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,563 @@
+            self._last_known_mode = current_mode
+
+        if (tgt := self._hk_target_mode(current_mode)) is not None:
+            self.char_target_state.set_value(tgt)
```
e32598aba13

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -1000,6 +1086,9 @@ def _async_create_single_accessory(
         conf = self._config.get(state.entity_id, {}).copy()
+        # Accessory mode has no aid allocation to tell new from existing,
+        # so the entity keeps the Thermostat and is offered the repair.
+        self._async_resolve_climate_type(state, conf, allow_auto=False)
```
a0325c74e11

bdraco commented on homeassistant/components/homekit/strings.json:
```diff
@@ -17,6 +17,19 @@
+      "fix_flow": {
+        "step": {
+          "confirm": {
+            "description": "The bridge {bridge} currently exposes {entity_id} as a Thermostat with a separate Fan. Switching to the Heater Cooler accessory puts the mode, temperature, fan speed, and swing on one tile that matches the device. The accessory keeps its room and name, but HomeKit scenes or automations that used the old Thermostat or Fan controls may need to be recreated. You can chang ...
```
a0325c74e11

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,576 @@
+                _LOGGER.debug(
+                    "%s: Ignoring off request; entity has no off mode",
+                    self.entity_id,
+                )
```
a0325c74e11

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,576 @@
+                service_calls.append(
+                    (SERVICE_SET_HVAC_MODE, {ATTR_HVAC_MODE: hass_mode})
+                )
+                self._last_known_mode = hass_mode
```
a0325c74e11

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -769,6 +820,46 @@ def add_bridge_accessory(self, state: State) -> HomeAccessory | None:
+            aid_storage.async_set_heater_cooler(
+                entity_id, climate_type == TYPE_HEATER_COOLER
+            )
+            return
```
a0325c74e11

bdraco commented on homeassistant/components/homekit/const.py:
```diff
@@ -260,6 +261,10 @@
+CHAR_CURRENT_HEATER_COOLER_STATE = "CurrentHeaterCoolerState"
+CHAR_TARGET_HEATER_COOLER_STATE = "TargetHeaterCoolerState"
+CHAR_COOLING_THRESHOLD_TEMPERATURE = "CoolingThresholdTemperature"
+CHAR_HEATING_THRESHOLD_TEMPERATURE = "HeatingThresholdTemperature"
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,542 @@
+                        break
+
+        if hasattr(self, "char_swing"):
+            current_swing = attributes.get(ATTR_SWING_MODE, "").lower()
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,542 @@
+                {ATTR_ENTITY_ID: self.entity_id, ATTR_FAN_MODE: fan_mode},
+            )
+        except (ServiceNotFound, ServiceValidationError) as e:
+            _LOGGER.error("Failed to set fan mode: %s", e)
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,542 @@
+        if self._supports_auto:
+            self._hk_to_ha_target[HC_TARGET_AUTO] = HVACMode.AUTO
+        else:
+            self._hk_to_ha_target[HC_TARGET_AUTO] = HVACMode.HEAT_COOL
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,542 @@
+        if CHAR_SWING_MODE in chars:
+            self.char_swing = serv.configure_char(CHAR_SWING_MODE, value=0)
+
+        self._last_known_mode: HVACMode = current_mode or HVACMode.COOL
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/const.py:
```diff
@@ -164,6 +164,7 @@
 SERV_TELEVISION = "Television"
 SERV_TELEVISION_SPEAKER = "TelevisionSpeaker"
 SERV_TEMPERATURE_SENSOR = "TemperatureSensor"
+SERV_HEATER_COOLER = "HeaterCooler"
```
4905e03b8b3

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1479 @@
+    await hass.async_block_till_done()
+
+    # With temp 22°C and target 20°C, current > target + delta (20.25)
+    # This should trigger: return HVACAction.COOLING in _derive_action (line 571)
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,532 @@
+            ):
+                selected_temp = cooling_temp
+            else:
+                selected_temp = heating_temp
```
4905e03b8b3

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1435 @@
+    # 72F current vs 68F target in cool mode is well past the 0.25C band
+    assert acc.char_current_state.value == HC_COOLING
+
+    # Within the hysteresis band (0.2F ~= 0.11C < 0.25C) stays idle
+    hass.states.async_set(
+        entity_id,
+        HVACMode.COOL,
+        {**base_attrs, ATTR_TEMPERATURE: 68.0, ATTR_CURRENT_TEMPERATURE: 68.2},
+    )
+    await hass.async_block_till_done()
+    assert acc.char_current_state.value == HC_IDLE
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,532 @@
+
+    @override
+    @callback
+    def async_update_state(self, new_state: State) -> None:
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,523 @@
+            if cur_c > tgt_c + ACTION_HYSTERESIS:
+                return HVACAction.COOLING
+            if cur_c < tgt_c - ACTION_HYSTERESIS:
+                return HVACAction.HEATING
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,523 @@
+        # Active/mode changes are handled first as they gate the others.
+        self._handle_active_mode_changes(char_values, service_calls)
+        self._handle_temperature_changes(char_values, service_calls)
+        self._handle_fan_swing_changes(char_values)
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,523 @@
+            ):
+                selected_temp = cooling_temp
+            else:
+                selected_temp = heating_temp
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,523 @@
+                return HVACAction.COOLING
+            if cur_c < tgt_c - ACTION_HYSTERESIS:
+                return HVACAction.HEATING
+        return HVACAction.IDLE
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,545 @@
+        return None
+
+    @override
+    @callback
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,545 @@
+                "climate",
+                service_name,
+                {ATTR_ENTITY_ID: self.entity_id, **service_data},
+            )
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_climate.py:
```diff
@@ -0,0 +1,58 @@
+"""Shared helpers for the climate and water heater accessory types."""
```
4905e03b8b3

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1487 @@
+
+
+async def test_heatercooler_basic(
+    hass: HomeAssistant, hk_driver: HomeDriver, events: list[Event]
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,536 @@
+        if supports_auto:
+            self._hk_to_ha_target[HC_TARGET_AUTO] = HVACMode.AUTO
+        elif supports_heat_cool:
+            self._hk_to_ha_target[HC_TARGET_AUTO] = HVACMode.HEAT_COOL
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_climate.py:
```diff
@@ -0,0 +1,86 @@
+"""Shared helpers for the climate and water heater accessory types."""
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,556 @@
+                CLIMATE_DOMAIN,
+                service_name,
+                {ATTR_ENTITY_ID: self.entity_id, **service_data},
+            )
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -150,7 +152,18 @@ def get_accessory(  # noqa: C901
+        # speeds or swing modes it can actually surface; otherwise the Thermostat
+        # handles it and existing accessories are preserved.
+        attributes = state.attributes
+        features = attributes.get(ATTR_SUPPORTED_FEATURES, 0)
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/climate_base.py:
```diff
@@ -0,0 +1,87 @@
+            CLIMATE_DOMAIN,
+            SERVICE_SET_SWING_MODE,
+            {ATTR_ENTITY_ID: self.entity_id, ATTR_SWING_MODE: mode},
+        )
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,500 @@
+
+        if temp_data:
+            service_calls.append((SERVICE_SET_TEMPERATURE, temp_data))
+
```
4905e03b8b3

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1672 @@
+    assert result is None
+
+    # Restore the mapping
+    acc._hk_to_ha_target = original_mapping
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/climate_util.py:
```diff
@@ -0,0 +1,153 @@
+"""Shared fan, swing, and temperature helpers for the climate accessory types."""
```
4905e03b8b3

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1702 @@
+    await hass.async_block_till_done()
+
+    service_calls = []
+    acc._handle_single_temp_changes(service_calls, 22.0, None)
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/climate_util.py:
```diff
@@ -0,0 +1,182 @@
+"""Shared fan, swing, and temperature helpers for the climate accessory types."""
```
4905e03b8b3

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1673 @@
+    await hass.async_block_till_done()
+
+    acc = HeaterCooler(hass, hk_driver, "HeaterCooler", entity_id, 2, None)
+    acc.run()
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,483 @@
+                current_temp_hk = self._temperature_to_homekit(current_temp)
+                if abs(cooling_temp - current_temp_hk) > abs(
+                    heating_temp - current_temp_hk
+                ):
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -150,7 +152,17 @@ def get_accessory(  # noqa: C901
+        has_swing = bool(features & ClimateEntityFeature.SWING_MODE) and (
+            get_swing_on_mode(attributes) is not None
+        )
+        a_type = "HeaterCooler" if has_fan or has_swing else "Thermostat"
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,483 @@
+                ):
+                    selected_temp = cooling_temp
+                else:
+                    selected_temp = heating_temp
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -150,7 +152,17 @@ def get_accessory(  # noqa: C901
+        has_swing = bool(features & ClimateEntityFeature.SWING_MODE) and (
+            get_swing_on_mode(attributes) is not None
+        )
+        a_type = "HeaterCooler" if has_fan or has_swing else "Thermostat"
```
4905e03b8b3

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -150,7 +155,19 @@ def get_accessory(  # noqa: C901
+            features & ClimateEntityFeature.FAN_MODE
+            or features & ClimateEntityFeature.SWING_MODE
+        ) and (
+            state.attributes.get(ATTR_FAN_MODES)
```
0583556feb4

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -150,7 +155,19 @@ def get_accessory(  # noqa: C901
+        ) and (
+            state.attributes.get(ATTR_FAN_MODES)
+            or state.attributes.get(ATTR_SWING_MODES)
+        ):
```
0583556feb4

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -150,7 +152,17 @@ def get_accessory(  # noqa: C901
+        has_swing = bool(features & ClimateEntityFeature.SWING_MODE) and (
+            get_swing_on_mode(attributes) is not None
+        )
+        a_type = "HeaterCooler" if has_fan or has_swing else "Thermostat"
```
0583556feb4

bdraco commented on tests/components/homekit/test_get_accessories.py:
```diff
@@ -455,6 +458,218 @@ def test_type_camera(type_name, entity_id, state, attrs) -> None:
+)
+def test_climate_accessory_selection(
+    type_name: str, entity_id: str, state: str, attrs: dict[str, Any]
+) -> None:
```
eab03b0ae2d

bdraco commented on tests/components/homekit/test_get_accessories.py:
```diff
@@ -455,6 +458,218 @@ def test_type_camera(type_name, entity_id, state, attrs) -> None:
+        ),
+    ],
+)
+def test_climate_accessory_type_override(
+    config_type: str, attrs: dict[str, Any], type_name: str
+) -> None:
```
eab03b0ae2d

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1875 @@
+    assert len(call_set_fan_mode) == 4
+    expected_calls = [FAN_LOW, FAN_MIDDLE, FAN_MEDIUM, FAN_HIGH]
+    for i, expected_mode in enumerate(expected_calls):
+        assert call_set_fan_mode[i].data[ATTR_FAN_MODE] == expected_mode
```
eab03b0ae2d

bdraco commented on homeassistant/components/homekit/type_heatercooler.py:
```diff
@@ -0,0 +1,534 @@
+        active_on = char_values.get(CHAR_ACTIVE) != 0
+        if active_on:
+            self._handle_temperature_changes(char_values, service_calls, current_state)
+
```
eab03b0ae2d

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,542 @@
+"""Class to hold all heater cooler accessories."""
```
9bdebf1992c

bdraco commented on tests/components/homekit/test_type_heatercooler.py:
```diff
@@ -0,0 +1,1652 @@
+    auto_attrs = {
+        ATTR_TEMPERATURE: 21.0,
+        ATTR_CURRENT_TEMPERATURE: 21.1,  # Within hysteresis range
+    }
```
ea7ef6d29b7

bdraco commented on homeassistant/components/homekit/config_flow.py:
```diff
@@ -537,10 +622,9 @@ async def async_step_include(
         if user_input is not None:
             entities = cv.ensure_list(user_input[CONF_ENTITIES])
             self.included_cameras = _async_cameras_from_entities(entities)
+            self.included_climates = _async_climates_from_entities(entities)
```
ea7ef6d29b7

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -769,6 +791,19 @@ def add_bridge_accessory(self, state: State) -> HomeAccessory | None:
+            and CONF_TYPE not in conf
+            and climate_supports_heater_cooler(state)
+        ):
+            self._heater_cooler_candidates.add(state.entity_id)
```
ea7ef6d29b7

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -151,7 +182,16 @@ def get_accessory(  # noqa: C901
+            # Only entities never bridged before route to the HeaterCooler
+            # automatically; existing ones keep their Thermostat and are offered
+            # the migration through a repair issue instead.
+            a_type = "HeaterCooler"
```
ea7ef6d29b7

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,560 @@
+        # drive one side. Any other mode (e.g. dry, fan_only) stays idle.
+        if mode in (HVACMode.HEAT_COOL, HVACMode.AUTO):
+            cool_above = attributes.get(ATTR_TARGET_TEMP_HIGH)
+            heat_below = attributes.get(ATTR_TARGET_TEMP_LOW)
```
ea7ef6d29b7

bdraco commented on homeassistant/components/homekit/config_flow.py:
```diff
@@ -536,11 +605,10 @@ async def async_step_include(
             entities = cv.ensure_list(user_input[CONF_ENTITIES])
-            self.included_cameras = _async_cameras_from_entities(entities)
+            self.included_cameras = _async_entities_in_domain(entities, CAMERA_DOMAIN)
+            self.included_climates = _async_entities_in_domain(entities, CLIMATE_DOMAIN)
```
ea7ef6d29b7

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -712,6 +749,8 @@ async def _async_recreate_removed_accessories_in_bridge_mode(
                 acc.run()
         self._async_update_accessories_hash()
+        # Recreated accessories can add new repair candidates.
+        self._async_update_heater_cooler_issues()
```
47ec2cc78e0

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -1000,6 +1070,10 @@ def _async_create_single_accessory(
+        if state.domain == CLIMATE_DOMAIN:
+            # Accessory mode has no aid allocation to tell new from existing,
+            # so the entity keeps the Thermostat and is offered the repair.
+            self._async_resolve_climate_type(state, conf, allow_auto=False)
```
47ec2cc78e0

bdraco commented on homeassistant/components/homekit/config_flow.py:
```diff
@@ -374,6 +378,69 @@ def __init__(self) -> None:
+            for label, entity_id in self._climate_choices.items():
+                entity_config = all_entity_config.setdefault(entity_id, {})
+
+                if (choice := user_input[label]) == CLIMATE_TYPE_AUTOMATIC:
+                    entity_config.pop(CONF_TYPE, None)
```
47ec2cc78e0

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,560 @@
+            selected_temp = cooling_temp
+        elif current_mode == HVACMode.HEAT:
+            selected_temp = heating_temp
+        elif (
+            current_mode == HVACMode.HEAT_COOL
+            and cooling_temp is not None
+            and heating_temp is not None
+        ):
```
e32598aba13

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,560 @@
+        attributes = new_state.attributes
+        current_mode = try_parse_enum(HVACMode, new_state.state)
+        if current_mode and current_mode != HVACMode.OFF:
+            self._last_known_mode = current_mode
```
e32598aba13

bdraco commented on homeassistant/components/homekit/aidmanager.py:
```diff
@@ -84,6 +86,16 @@ async def async_initialize(self) -> None:
+        so it is written down to survive restarts.
+        """
+        self.heater_cooler_entities.add(entity_id)
+        self.async_schedule_save()
```
e32598aba13

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -769,6 +812,33 @@ def add_bridge_accessory(self, state: State) -> HomeAccessory | None:
+        if entity_id in aid_storage.heater_cooler_entities:
+            conf[CONF_TYPE] = TYPE_HEATER_COOLER
+            return
+        if not climate_supports_heater_cooler(state):
```
e32598aba13

bdraco commented on tests/components/homekit/test_aidmanager.py:
```diff
@@ -647,3 +647,62 @@ async def test_handle_unique_id_change(
+    with patch(
+        "homeassistant.components.homekit.aidmanager.AccessoryAidStorage.async_schedule_save"
+    ):
+        aid_storage = AccessoryAidStorage(hass, config_entry)
```
e32598aba13

bdraco commented on homeassistant/components/homekit/config_flow.py:
```diff
@@ -374,6 +389,77 @@ def __init__(self) -> None:
+                if (choice := user_input[label]) == CLIMATE_TYPE_AUTOMATIC:
+                    entity_config.pop(CONF_TYPE, None)
+                else:
+                    entity_config[CONF_TYPE] = choice
```
home-assistant/home-assistant.io@b2cb5b178d

bdraco committed:
Store a generic accessory type and fix category and fan edge cases

AccessoryAidStorage now stores an opaque accessory type per entity
instead of knowing about heater coolers, standalone pairings
advertise the heater or air conditioner category, and the linked
fan starts inactive for an unavailable entity

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,573 @@
+
+    def __init__(self, *args: Any) -> None:
+        """Initialize a HeaterCooler accessory object."""
+        super().__init__(*args)
```
004e6636162

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,573 @@
+            (humidity := attributes.get(ATTR_CURRENT_HUMIDITY)), (int, float)
+        ):
+            self.char_current_humidity.set_value(humidity)
+        if self.fan_chars:
+            self._async_update_fan_service(new_state)
```
004e6636162

bdraco committed:
Simplify category selection, inactive states, and write rejection

The category derives from the existing can_cool flag, the inactive
state set lives in one constant, rejected client writes share a
revert helper, the first start flag clears once the pairing is
persisted, and the repairs tests lose the module level setup wrapper

bdraco committed:
Merge branch 'dev' into heatercooler-support

bdraco committed:
Record the automatic routing only after the accessory is created

The resolver returns the pending choice and the caller commits it
once the accessory exists, so a construction failure cannot wedge
an entity on a type that never produced an accessory

bdraco committed:
Merge remote-tracking branch 'nitaybz/heatercooler-support' into heatercooler-support

bdraco committed:
Move climate type resolution into its own module

HomeKit no longer tracks repair candidates or issue ids directly;
ClimateTypeResolver owns the routing decision, the candidacy
bookkeeping, and the issue lifecycle

bdraco committed:
Move the accessory type resolver into accessories

accessories.py already owns type determination through get_accessory,
so the resolver, candidacy bookkeeping, and issue lifecycle live
there now and HomeKit only holds a generic AccessoryTypeResolver

bdraco committed:
Restore unsupported target writes for dry only entities

The restore falls back to the default target when the last known
mode has no HomeKit representation, and the candidate issue constant
follows the generic accessory type naming

bdraco committed:
Defer the explicit type persistence like the automatic one

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,572 @@
+            elif (restore := self._hk_target_mode(self._last_known_mode)) is not None:
+                # The write already changed the characteristic to a target
+                # the entity cannot enter, so put it back on the last mode.
+                self._reject_char_write(self.char_target_state, restore)
```
7b9a83082f2

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -151,7 +185,8 @@ def get_accessory(  # noqa: C901
         a_type = "BinarySensor"
 
     elif state.domain == "climate":
-        a_type = "Thermostat"
+        # The type is resolved by the bridge before the accessory is created.
+        a_type = CLIMATE_TYPES[config.get(CONF_TYPE, TYPE_THERMOSTAT)]
 
```
923f9ec2ea9

bdraco committed:
Pick range or single setpoint writes from the effective mode

Dual capable entities publish the range keys even in heat or cool
mode, so writes follow the requested or current mode and the display
follows the reported values instead of key presence

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,572 @@
+        supports_dual_temp = current_state is not None and (
+            ATTR_TARGET_TEMP_HIGH in current_state.attributes
+            or ATTR_TARGET_TEMP_LOW in current_state.attributes
+        )
```
3aa28696d96

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,572 @@
+        """Update HomeKit temperature thresholds based on HA state."""
+        if not self._has_cool_threshold and not self._has_heat_threshold:
+            return
+        attributes = state.attributes
+        supports_dual_temp = (
+            ATTR_TARGET_TEMP_HIGH in attributes or ATTR_TARGET_TEMP_LOW in attributes
+        )
+
+        if supports_dual_temp:
```
3aa28696d96

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,577 @@
+        supports_dual_temp = current_state is not None and (
+            ATTR_TARGET_TEMP_HIGH in current_state.attributes
+            or ATTR_TARGET_TEMP_LOW in current_state.attributes
+        )
```
3aa28696d96

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,577 @@
+        """Update HomeKit temperature thresholds based on HA state."""
+        if not self._has_cool_threshold and not self._has_heat_threshold:
+            return
+        attributes = state.attributes
+        supports_dual_temp = (
+            ATTR_TARGET_TEMP_HIGH in attributes or ATTR_TARGET_TEMP_LOW in attributes
+        )
+
+        if supports_dual_temp:
```
3aa28696d96

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -126,6 +143,176 @@
+        self._candidates.pop(entity_id, None)
+        if climate_type := conf.get(CONF_TYPE):
+            aid_storage.async_set_accessory_type(entity_id, climate_type)
+            return None
```
780e371b84e

bdraco committed:
Simplify mode resolution and pending type recording

The batched target mode is extracted once per write and the
effective mode resolved once, the target mapping gets a reverse
lookup so fallback modes are representable and the default target
field goes away, both creation sites share a recording helper, and
the issue id helper drops its stale async prefix and wording

bdraco committed:
Apply climate writes in order and roll back failed allocations

Mode and temperature writes are awaited sequentially before the fan
and swing changes so a setpoint cannot race its mode switch, a
failed first creation rolls back the new aid allocation so the
entity is retried as new, and a range write requires both threshold
characteristics so contradictory device configs degrade to a single
setpoint

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -753,22 +770,37 @@ def add_bridge_accessory(self, state: State) -> HomeAccessory | None:
+        pending_type = self._type_resolver.async_resolve(
+            self.aid_storage, state, conf, allow_auto=True
+        )
+        aid = self.aid_storage.get_or_allocate_aid_for_entity_id(state.entity_id)
```
e0617fb3615

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,593 @@
+                CLIMATE_DOMAIN,
+                service_name,
+                {ATTR_ENTITY_ID: self.entity_id, **service_data},
+            )
```
e0617fb3615

bdraco committed:
Abort a write batch on failure and fix the fan and dry only category

A rejected mode write stops the remaining temperature, fan, and
swing writes since the tile was already re-synced, and entities that
neither heat nor cool advertise the air conditioner category

bdraco committed:
Drop the repair based migration

With the options flow in place the repair issues are redundant;
existing entities keep their Thermostat and users switch to the
Heater Cooler at their own pace, so the repairs platform, the issue
lifecycle, the candidate bookkeeping, and the resolver class go
away in favor of a stateless resolve function

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,615 @@
+
+        # Standalone pairings advertise the category in the QR code and mDNS
+        # metadata, so pick the one matching the device instead of Thermostat.
+        self.category = CATEGORY_AIR_CONDITIONER if can_cool else CATEGORY_HEATER
```
7fe68ef6e8c

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,615 @@
+                CLIMATE_DOMAIN,
+                service_name,
+                {ATTR_ENTITY_ID: self.entity_id, **service_data},
+            )
```
7fe68ef6e8c

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -126,6 +143,178 @@
+                accessory_type_issue_id(self._entry_id, entity_id),
+                is_fixable=True,
+                severity=ir.IssueSeverity.WARNING,
+                translation_key=ISSUE_ACCESSORY_TYPE_CANDIDATE,
```
e4124baa8c7

bdraco commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -1079,8 +1169,42 @@ async def _async_create_accessories(self) -> bool:
+                DOMAIN,
+                _heater_cooler_issue_id(self._entry_id, entity_id),
+                is_fixable=True,
+                severity=ir.IssueSeverity.WARNING,
```
e4124baa8c7

bdraco committed:
Serialize write batches and require an advertised swing off mode

Batches take a per accessory lock so rapid writes cannot interleave,
automatic routing requires the entity to advertise a swing off mode
since the binary control writes it back, the recording helper takes
the narrowed aid storage instead of re-asserting, and the event fire
and resync failures are logged by the component instead of the
loop's task handler

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -126,6 +143,178 @@
+    has_fan = bool(features & ClimateEntityFeature.FAN_MODE) and (
+        len(get_fan_modes_and_speeds(attributes)[1]) >= 2
+    )
+    has_swing = bool(features & ClimateEntityFeature.SWING_MODE) and (
+        get_swing_on_mode(attributes) is not None
+    )
```
83ff3d91ca9

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,625 @@
+                    service_calls, char_values if active_on else None
+                ),
+                eager_start=True,
+            )
```
83ff3d91ca9

bdraco committed:
Back the Auto target with heat cool and commit accepted modes only

HEAT_COOL keeps its thresholds adjustable while AUTO may follow a
schedule, matching the thermostat precedence, and the remembered
mode is committed only after the entity accepts the write so a
rejected mode is never restored

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,625 @@
+        if supports_auto:
+            self._hk_to_ha_target[HC_TARGET_AUTO] = HVACMode.AUTO
+        elif supports_heat_cool:
+            self._hk_to_ha_target[HC_TARGET_AUTO] = HVACMode.HEAT_COOL
```
9d9ecff902f

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,625 @@
+                service_calls.append(
+                    (SERVICE_SET_HVAC_MODE, {ATTR_HVAC_MODE: requested_mode})
+                )
+                self._last_known_mode = requested_mode
```
9d9ecff902f

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,625 @@
+                self._async_apply_writes(
+                    service_calls, char_values if active_on else None
+                ),
+                eager_start=True,
```
83ff3d91ca9

bdraco committed:
Queue fan and swing writes under the batch lock

Fan and swing changes join the ordered write queue so no batch can
bypass or outrun the lock, a target bundled with an off write is
remembered for the next power on or put back when unsupported, and
the driver test parameters gain their type

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,631 @@
+                eager_start=True,
+            )
+        elif active_on:
+            self._handle_fan_swing_changes(char_values)
```
f7c1ea83708

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,631 @@
+        if active is None and target_mode is None:
+            return
+
+        if active == 0:
```
f7c1ea83708

bdraco commented on tests/components/homekit/test_accessories.py:
```diff
@@ -768,6 +768,30 @@ async def test_call_service_resyncs_on_failure(
 
 
+async def test_call_service_resync_failure_is_logged(
+    hass: HomeAssistant, hk_driver, caplog: pytest.LogCaptureFixture
```
f7c1ea83708

bdraco committed:
Keep the service call try block scoped to the call

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,638 @@
+                ):
+                    self._last_known_mode = mode
+            if fan_swing_char_values:
+                self._handle_fan_swing_changes(fan_swing_char_values)
```
f7c1ea83708

bdraco commented on tests/components/homekit/test_accessories.py:
```diff
@@ -768,6 +768,30 @@ async def test_call_service_resyncs_on_failure(
     assert pushed_state.state == STATE_OFF
 
 
+async def test_call_service_resync_failure_is_logged(
+    hass: HomeAssistant, hk_driver, caplog: pytest.LogCaptureFixture
+) -> None:
```
f7c1ea83708

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -645,38 +768,36 @@ def async_call_service(
-            else:
-                return
+        try:
+            self.hass.bus.async_fire(EVENT_HOMEKIT_CHANGED, event_data, context=context)
+            await self.hass.services.async_call(
+                domain, service, service_data, blocking=True, context=context
+            )
```
18580e026d9

bdraco committed:
Resolve write batches under the accessory lock

The mode and setpoint decisions now run inside the locked coroutine
so a batch sees the outcome of the one before it instead of the
stale pre-switch state

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,654 @@
+        self._handle_active_mode_changes(
+            char_values, service_calls, current_state, requested_mode
+        )
+        # Temperature and fan/swing writes are meaningless when turning off.
+        active_on = char_values.get(CHAR_ACTIVE) != 0
+        if active_on:
+            self._handle_temperature_changes(
+                char_values, service_calls, current_state, requested_mode
+            )
```
d40bb6e7df8

bdraco committed:
Resolve power on setpoints against the mode being activated

Thresholds batched with an Active on write use the last known mode
as the effective mode so dual capable entities get a range write
instead of a single setpoint picked from the off state

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,651 @@
+                    char_values, service_calls, current_state, requested_mode
+                )
+                # Fan and swing are queued last so they follow the mode switch.
+                self._queue_fan_swing_changes(char_values, service_calls)
```
3f1074afce0

bdraco committed:
Simplify mode memory and route every climate write through the lock

Queued writes carry the mode to remember once accepted, collapsing
the service name matching and the bundled target special case into
one commit rule, the unsupported target restore lives in one guard,
linked fan service writes go through a locked dispatch so no write
bypasses the batch ordering, and the one line recording helper is
inlined at its call sites

bdraco committed:
Keep an accepted mode effective until the entity reports it

Push integrations can return from set_hvac_mode before their state
callback arrives, so the accepted mode bridges following batches
and clears once the entity reports a mode again

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,655 @@
+        """
+        async with self._write_lock:
+            service_calls: list[tuple[str, dict[str, Any]]] = []
+            current_state = self.hass.states.get(self.entity_id)
```
aa39009d013

bdraco committed:
Share the whole domain include expansion between cameras and climate

A camera domain included without explicit entities now offers the
camera options step, matching the climate behavior

bdraco committed:
Fix write pipeline and routing edge cases found in review

Repeated off writes can no longer commit off into the mode memory,
a quick off to on toggle sends the mode again while the off is
still pending, a rejected off keeps the rest of the batch applying,
a pending mode no longer masks the unsupported target rejection,
range only entities with one sided modes expose both thresholds so
their writes stay valid, the reset service looks up aids without
allocating so unbridged entities stay new, stored accessory types
heal to the current identity on read so repeated unique id changes
cannot orphan them, and the HomeKit temperature bounds round inward
so slider extremes cannot produce writes the entity rejects; the
write lock moves to a decorator

bdraco committed:
Keep the pending mode until the entity reports a mode change

Attribute updates that re-report the pre-switch mode no longer clear
the bridge, so threshold batches keep resolving against the accepted
mode until the entity actually moves

bdraco committed:
Simplify the batch policy and identity lookups

The active mode handler owns whether the batch continues, queue
entries declare the pending and remembered modes they apply once
accepted instead of the loop matching service names, and the aid
allocation check reuses the non allocating lookup

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,687 @@
+                # A just accepted mode stays effective until the entity
+                # reports it, so a following batch does not resolve against
+                # the pre-switch state.
+                requested_mode = self._pending_mode
```
2719e76a6ab

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,687 @@
+        current_mode = try_parse_enum(HVACMode, new_state.state)
+        if current_mode is not None:
+            # The entity reported a mode, so its state is authoritative again
+            self._pending_mode = None
```
5aa38726328

bdraco committed:
Keep exact limits when inward rounding would cross over

A range too narrow to hold a half degree step keeps the entity's
exact limits instead of being swapped outward by the reversed range
handling

bdraco commented on homeassistant/components/homekit/climate_util.py:
```diff
@@ -125,12 +132,14 @@ def get_temperature_range_from_state(
-        min_temp = round(temperature_to_homekit(min_temp, unit) * 2) / 2
+        # Round inward so the slider cannot produce a write the entity's
+        # own limit validation rejects
+        min_temp = math.ceil(temperature_to_homekit(min_temp, unit) * 2) / 2
```
994a2cbe45b

bdraco commented on homeassistant/components/homekit/climate_util.py:
```diff
@@ -125,12 +132,14 @@ def get_temperature_range_from_state(
 
     if (max_temp := state.attributes.get(ATTR_MAX_TEMP)) is not None:
-        max_temp = round(temperature_to_homekit(max_temp, unit) * 2) / 2
+        max_temp = math.floor(temperature_to_homekit(max_temp, unit) * 2) / 2
```
994a2cbe45b

bdraco committed:
Log when a failed write cannot re-sync a missing entity

bdraco committed:
Let a fresh state report win over queued mode values

A blocking service can publish its state before the call returns,
so the queued pending and remembered modes only apply when their
counterparts were not updated while awaiting, the temperature
bounds round inward to the characteristic's 0.1 step so valid
setpoints near the limits stay selectable, and the added divider
comment goes away

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,721 @@
+            if call.pending_mode:
+                self._pending_mode = call.pending_mode
+            if call.commit_mode:
+                self._last_known_mode = call.commit_mode
```
bad5d638678

bdraco commented on homeassistant/components/homekit/const.py:
```diff
@@ -128,6 +128,10 @@
 TYPE_FAN = "fan"
 TYPE_AIR_PURIFIER = "air_purifier"
 
+# #### Climate Types ####
```
bad5d638678

bdraco commented on homeassistant/components/homekit/climate_util.py:
```diff
@@ -125,18 +132,26 @@ def get_temperature_range_from_state(
+    # limit validation rejects; a range too narrow to hold a half degree
+    # step keeps the exact limits rather than expanding beyond them.
+    rounded_min = math.ceil(min_temp * 2) / 2
+    rounded_max = math.floor(max_temp * 2) / 2
```
bad5d638678

bdraco committed:
Show the current accessory in the climate options and honor pending state

The climate step labels the entities with the accessory they use
right now so Automatic is not a mystery, a stale re-report of the
pre-switch mode no longer flips the tile or the restore target off
the pending mode, and the swing toggle is only exposed when the
entity advertises an off mode since the off write would otherwise
be rejected

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,725 @@
+            self._last_reported_mode = current_mode
+        if current_mode and (tgt := self._hk_target_mode(current_mode)) is not None:
+            self._last_known_mode = current_mode
+            self.char_target_state.set_value(tgt)
```
f3f468847df

bdraco commented on homeassistant/components/homekit/type_heater_coolers.py:
```diff
@@ -0,0 +1,725 @@
+        # characteristics when the entity exposes predefined modes.
+        if self.ordered_fan_speeds and not self.fan_chars:
+            chars.append(CHAR_ROTATION_SPEED)
+        if self.swing_on_mode is not None:
+            chars.append(CHAR_SWING_MODE)
```
f3f468847df

bdraco committed:
Give every service an explicit primary flag and name the humidity sensor

The primary service is set after the fan and humidity services exist
so none are left without a flag for the Home app to guess at, and
the linked humidity sensor carries a configured name

bdraco committed:
Mark the accessory type read as loop bound since it can persist

bdraco commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -126,6 +142,94 @@
+        # one, so every path that sets a type defers persistence until
+        # the accessory exists.
+        return cast(str, climate_type)
+    if aid_storage.get_accessory_type(entity_id) == TYPE_HEATER_COOLER:
+        if not climate_controls_target_humidity(state):
+            conf[CONF_TYPE] = TYPE_HEATER_COOLER
+            return None
+        # A humidity setpoint gained since the choice was stored cannot
+        # be represented by the HeaterCooler, so the routing is dropped.
+        aid_storage.async_set_accessory_type(entity_id, None)
```
ac912afcdd7 (#176229)

bdraco commented on tests/components/homekit/test_config_flow.py:
```diff
@@ -1611,3 +1655,200 @@ async def test_options_flow_include_mode_allows_hidden_entities(
+
+    # A loaded entry exposes the bridged accessories through its runtime
+    # data; accessory mode reads the single accessory from the driver
+    thermostat = type("Thermostat", (), {"entity_id": "climate.new"})()
+    if homekit_mode == "bridge":
+        homekit = Mock(bridge=Mock(accessories={2: thermostat}))
+    else:
+        homekit = Mock(bridge=None, driver=Mock(accessory=thermostat))
```
ac912afcdd7 (#176229)
